# Bệnh đái tháo đường ở nữ giới Pima Indian

## 1. Định nghĩa vấn đề (Define Problem)

* Phân tích vấn đề:
- Người da đỏ Pima tại Arizona có tỷ lệ mắc tiểu đường type 2 rất cao.
- Viện Quốc gia về Bệnh tiểu đường, Tiêu hóa và Thận (NIDDK) đã nghiên cứu liên tục từ 1965.
- Bệnh tiểu đường được chẩn đoán theo WHO: glucose huyết tương ≥ 200 mg/dL sau 2 giờ OGTT hoặc được phát hiện trong chăm sóc y tế thường quy.
- Mục tiêu: xây dựng mô hình dự báo khả năng mắc bệnh tiểu đường trong vòng 5 năm, dựa trên các yếu tố nguy cơ.

* Mô tả: 
 Dataset Pima Indians Diabetes chứa thông tin của 768 bệnh nhân nữ, ít nhất 21 tuổi, thuộc cộng đồng Pima Indian gần Phoenix, Arizona, USA. Mục tiêu là dự đoán xem một bệnh nhân có bị đái tháo đường (Class = 1) hay không (Class = 0) dựa trên 8 thuộc tính y tế.

* Dữ liệu vào:
   1. Pregnancies: Số lần mang thai, phản ánh tiền sử sinh sản có thể liên quan đến nguy cơ đái tháo đường.
   2. Glucose: Nồng độ glucose huyết tương sau 2 giờ trong bài kiểm tra dung nạp glucose (mg/dl), chỉ số quan trọng để chẩn đoán đái tháo đường.
   3. BloodPressure: Huyết áp tâm thu (mm Hg), đo áp lực máu khi tim giãn ra, liên quan đến sức khỏe tim mạch.
   4. SkinThickness: Độ dày nếp gấp da tay (mm), đại diện cho lượng mỡ dưới da, một yếu tố liên quan đến béo phì.
   5. Insulin: Mức insulin huyết thanh sau 2 giờ (mu U/ml), phản ánh khả năng sản xuất insulin của cơ thể.
   6. BMI: Chỉ số khối cơ thể (kg/m²), đo lường mức độ béo phì, một yếu tố nguy cơ lớn của đái tháo đường loại 2.
   7. DiabetesPedigreeFunction: Hàm phả hệ đái tháo đường, đánh giá nguy cơ di truyền dựa trên lịch sử gia đình.
   8. Age: Tuổi (năm), ảnh hưởng đến nguy cơ mắc bệnh do sự lão hóa.
   9. Outcome: Biến lớp (0: Không đái tháo đường, 1: Có đái tháo đường), mục tiêu chính của phân tích.
    
    - Trong đó  Hàm phả hệ đái tháo đường(DPF):
        $$
        + DPF = \frac{\sum_i K_i (88 - ADM_i) + 20}{\sum_j K_j (ALC_j - 14) + 50}
        $$
        + Được xây dựng để lượng hóa yếu tố di truyền dựa trên tiền sử bệnh tiểu đường của cha mẹ, ông bà, anh chị em, họ hàng gần.
        + DPF tăng khi số lượng thân nhân mắc bệnh tăng, tuổi mắc bệnh trẻ hơn, và mức độ quan hệ huyết thống gần.
        + DPF giảm khi thân nhân không mắc bệnh và có tuổi thọ cao.

* Lựa chọn trường hợp:
    - Tiêu chí:
        + Nữ, ≥ 21 tuổi, OGTT ban đầu không bị tiểu đường.
    - Sau 5 năm:
        + Hoặc phát hiện tiểu đường (positive case).
        + Hoặc kiểm tra lại ≥ 5 năm sau vẫn không mắc (negative case).
        + Loại bỏ trường hợp phát hiện tiểu đường trong vòng < 1 năm (do quá dễ dự báo).
    - Tổng số mẫu: 768
        + 576 mẫu cho tập huấn luyện
        + 192 mẫu cho tập dự báo (test set).

* Hướng giải quyết:
    - Sử dụng ADAP (Adaptive Learning Algorithm) – một dạng mô hình học máy dựa trên perceptron, để dự báo khả năng mắc bệnh trong vòng 5 năm.
    - Biến đầu vào gồm 8 yếu tố nguy cơ chính + nhãn phân loại (có/không mắc bệnh).

* Kết quả (từ nghiên cứu trước đó [Smith et al., 1988]):
    - ADAP đưa ra dự đoán dạng xác suất (0–1), sau đó quy đổi thành nhị phân.
    - Với cutoff 0.448, mô hình đạt:
        + Sensitivity (độ nhạy): ~76%
        + Specificity (độ đặc hiệu): ~76%
    - Đây là kết quả khá tốt trong bối cảnh dữ liệu y tế thời kỳ đó, và bộ dữ liệu trở thành benchmark nổi tiếng trong cộng đồng Machine Learning.

* Kết quả: 
    - Biến mục tiêu (Outcome): Biến nhị phân (0 hoặc 1) biểu thị trạng thái đái tháo đường
        + 0: Không bị đái tháo đường (500 mẫu).
        + 1: Bị đái tháo đường (268 mẫu).
* Thông tin bổ sung:
    - Giá trị thiếu: Dữ liệu có chứa giá trị thiếu, thường được biểu thị bằng 0 trong các cột như Glucose, BloodPressure, SkinThickness, Insulin, và BMI, cần được xử lý trong quá trình phân tích.
    - Nghiên cứu trước: Một nghiên cứu trước đây của Smith et al. (1988) sử dụng thuật toán ADAP, đạt độ nhạy (sensitivity) và độ đặc hiệu (specificity) 76% trên 192 mẫu kiểm tra sau khi huấn luyện với 576 mẫu, với ngưỡng cắt 0.448.
    - Mục tiêu phân tích: Xây dựng một mô hình dự đoán hiệu quả, so sánh với kết quả ADAP, và khám phá các yếu tố quan trọng ảnh hưởng đến nguy cơ đái tháo đường.

## 2. Chuẩn bị vấn đề (Prepare Problem)

### 2.1 Mô tả vấn đề
- Bệnh tiểu đường type 2 có tỷ lệ rất cao trong cộng đồng này.
- Việc chẩn đoán chỉ dựa vào tiêu chuẩn WHO OGTT (≥ 200 mg/dL sau 2 giờ) mới cho kết luận rõ ràng, nhưng điều này chỉ xác định khi bệnh đã xuất hiện.
- Mục tiêu:
    +  Trong phần này, chúng ta sẽ thiết lập môi trường làm việc bằng cách khai báo các thư viện cần thiết, load dataset Pima Indians Diabetes, và giải thích ý nghĩa của từng cột dữ liệu để bạn dễ dàng hiểu rõ giá trị mà chúng mang lại. Đây là bước nền tảng quan trọng để tiến hành phân tích sâu hơn về nguy cơ đái tháo đường trong cộng đồng Pima Indian.
    + Nhu cầu đặt ra: dự đoán sớm nguy cơ mắc bệnh tiểu đường trong tương lai dựa trên các yếu tố nguy cơ (tuổi, BMI, insulin, huyết áp, di truyền…).
- Dữ liệu:
    -  Dataset được cung cấp dưới dạng file pima-indians-diabetes.csv, chứa 768 mẫu với 8 thuộc tính số và 1 biến lớp (Outcome). Dữ liệu có thể chứa giá trị thiếu, cần được xử lý phù hợp. 
- Thách thức:
    + Dữ liệu y tế thường có mối quan hệ phi tuyến phức tạp giữa các biến.
    + Có sự hiện diện của missing values và outliers (ví dụ insulin = 0).
    + Cần một mô hình dự đoán chính xác nhưng cũng có thể giải thích được trong ngữ cảnh y tế.

### 2.1 Khai báo thư viện (Load Libraries)

- **pandas** là một thư viện mạnh mẽ trong Python, được sử dụng rộng rãi cho **phân tích và xử lý dữ liệu**.
- Cung cấp cấu trúc dữ liệu chính:
  - **Series**: tương tự như một mảng 1 chiều có nhãn.
  - **DataFrame**: bảng dữ liệu 2 chiều (giống Excel hoặc SQL table).
- Các tính năng chính:
  - Đọc/ghi dữ liệu từ nhiều định dạng (CSV, Excel, SQL, JSON, ...).
  - Xử lý dữ liệu: lọc, nhóm, gộp, sắp xếp, thống kê.
  - Làm sạch dữ liệu: xử lý missing values, outliers.
  - Hỗ trợ tích hợp với NumPy, Matplotlib, Seaborn để trực quan hóa dữ liệu.

In [3]:
import pandas as pd

### 2.2 Nạp dữ liệu (Load Dataset)

- Thực hiện việc **nạp dữ liệu gốc** từ file `pima-indians-diabetes.csv` vào Python bằng thư viện `pandas`.  
- Do file CSV không chứa tiêu đề cột (header), ta cần **tự gán tên cột** dựa trên mô tả trong tài liệu nghiên cứu gốc.  
- Kết quả: dữ liệu được lưu vào một **DataFrame** có cấu trúc rõ ràng, giúp dễ dàng phân tích, trực quan hóa và xây dựng mô hình dự báo.

In [4]:
# Đọc dữ liệu từ file CSV
data_path = 'pima-indians-diabetes.csv'  # Đường dẫn tới file dữ liệu

# Gán tên cột dựa trên mô tả trong tài liệu
column_names = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'] # Tên các cột

# Đọc dữ liệu từ file CSV
data = pd.read_csv(data_path, header=None, names=column_names)

### 2.3 Kiểm tra một số thông tin cơ bản

- Sử dụng các hàm cơ bản của pd đễ hỗ trợ cho chúng ta có một cái nhìn tổng quát về số liệu của file 

In [5]:
# Hiển thị 5 dòng đầu để kiểm tra
print("5 dòng đầu của dữ liệu:")
print(data.head())

5 dòng đầu của dữ liệu:
   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  


In [6]:
(data==0).sum()

Pregnancies                 111
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                     500
dtype: int64

- **Nhận xét về dữ liệu có giá trị 0:**  
   1. Pregnancies (111 giá trị 0):
      + 0 lần mang thai là hợp lý (có những phụ nữ chưa từng mang thai).
      + Không cần xử lý coi là missing.
   2. Glucose (5 giá trị 0):
      + Về mặt y khoa, glucose máu không thể = 0.
      + Đây là giá trị bất thường → coi là dữ liệu thiếu.
   3. BloodPressure (35 giá trị 0):
      + Huyết áp tâm trương không thể = 0.
      + Các giá trị này là missing values cần xử lý
   4. SkinThickness (227 giá trị 0):
      + 0 mm nếp gấp da là không hợp lý.
      + Đây là một biến có nhiều thiếu dữ liệu (≈ 30% dữ liệu).
   5. Insulin (374 giá trị 0):
      + 0 µU/mL insulin là không thực tế.
      + Biến này có nhiều dữ liệu thiếu nhất (≈ 49%).
   6. BMI (11 giá trị 0):
      + BMI = 0 là bất hợp lý.
      + Một số ít dữ liệu cần xử lý missing.
   7. DiabetesPedigreeFunction (0 giá trị 0):
      + Không có giá trị 0, dữ liệu đầy đủ.
   8. Age (0 giá trị 0):
      + Không có giá trị 0, dữ liệu đầy đủ.
   9. Outcome (500 giá trị 0):
      + Đây là biến nhãn (class):
         + 0 = không mắc tiểu đường (500 mẫu).
         + 1 = mắc tiểu đường (268 mẫu).
      + Đây không phải dữ liệu thiếu mà là nhãn hợp lệ.

In [7]:
# Hiển thị thông tin cơ bản về dữ liệu
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB
None


- Nhận xét sơ bộ
- **Cấu trúc dữ liệu:**  
   - Tổng cộng **767 dòng × 9 cột**.  
   - Tất cả các biến đều dạng số (`int` hoặc `float`). 

In [8]:
# Thống kê mô tả
print(data.describe())


       Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  

- Các thống kê mô tả chính và ý nghĩa:
    + count: số lượng mẫu dữ liệu (ở đây tất cả cột = 768).
    + mean: giá trị trung bình.
    + std: độ lệch chuẩn → mức độ phân tán quanh trung bình.
    + min, max: giá trị nhỏ nhất và lớn nhất.
    + 25%, 50%, 75% (quartiles): các mốc phân vị → dùng để đánh giá sự phân phối dữ liệu.

- Chỉ số đo lường sự biến động (Dispersion):
    + std (độ lệch chuẩn):
        + Glucose: 32 → biến thiên khá lớn, thể hiện sự khác biệt giữa người khỏe mạnh và người có nguy cơ cao.
        + Insulin: 115, lớn hơn cả mean (~80) → chứng tỏ dữ liệu rất phân tán và nhiều ngoại lệ.
        + BMI: 7.88 → tương đối ổn định, dao động vừa phải.
        + Age: 11.8 → đa dạng độ tuổi, từ 21 đến 81 tuổi.
- Khoảng (range = max - min):
    + Pregnancies: 0–17, cho thấy có vài trường hợp đặc biệt mang thai rất nhiều lần.
    + Insulin: 0–846, cực kỳ phân tán → cần chuẩn hóa hoặc log-transform khi phân tích.
    + SkinThickness: 0–99, cũng có nhiều giá trị bất hợp lý.

In [9]:
# Phân bố lớp
print("\nPhân bố biến lớp (Outcome):")
print(data['Outcome'].value_counts())


Phân bố biến lớp (Outcome):
Outcome
0    500
1    268
Name: count, dtype: int64


- Tỉ lệ mắc ~35% và không mắc ~65%

## 3. Chuẩn bị dữ liệu (Prepare Data)

### 3.1 Mô tả vấn đề

- Mục tiêu: Trong phần này, chúng ta sẽ làm sạch dataset Pima Indians Diabetes bằng cách xử lý giá trị thiếu, loại bỏ dữ liệu trùng lặp, và thực hiện các thao tác cần thiết để đảm bảo dữ liệu sẵn sàng cho phân tích sâu hơn. Điều này giúp loại bỏ các sai lệch và chuẩn bị một nền tảng vững chắc cho các bước tiếp theo như trực quan hóa và mô hình hóa.

### 3.1 Làm sạch dữ liệu (Data Cleaning)

#### Xử lý dữ liệu bị thiếu

- ***Giới thiệu thư viện NumPy và SimpleImputer***
- **NumPy (`numpy`)**:
    + Là thư viện **tính toán khoa học** cốt lõi trong Python.  
    + Hỗ trợ mảng đa chiều (`ndarray`) và nhiều hàm toán học tối ưu.  
    + Trong phân tích dữ liệu, NumPy thường được dùng kết hợp với `pandas` để xử lý dữ liệu số học, thống kê, và tối ưu hiệu năng. 
- **SimpleImputer (từ `sklearn.impute`)**:
    + Là công cụ trong thư viện **Scikit-learn** hỗ trợ **xử lý dữ liệu thiếu (missing values)**.  
    + Có thể thay thế (impute) giá trị thiếu bằng:  
        - Trung bình (`mean`)  
        - Trung vị (`median`)  
        - Giá trị thường gặp nhất (`most_frequent`)  
        - Một hằng số tùy chọn (`constant`) 
    + Đây là bước quan trọng để đảm bảo dữ liệu **sạch và đầy đủ** trước khi đưa vào huấn luyện mô hình học máy.   

In [10]:
import numpy as np
from sklearn.impute import SimpleImputer

ModuleNotFoundError: No module named 'sklearn'

In [ ]:
# Thay thế giá trị 0 bằng NaN để xử lý
for column in ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']:
    data[column] = data[column].replace(0, np.nan)

# Sử dụng SimpleImputer để điền giá trị thiếu bằng trung bình
imputer = SimpleImputer(strategy='mean')
data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']] = imputer.fit_transform(data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']])

# Kiểm tra lại số lượng giá trị thiếu
print("Số lượng giá trị thiếu sau khi xử lý:")
print(data.isnull().sum())

- Giải thích:

    - Việc thay thế bằng trung bình giúp duy trì tính đại diện của dữ liệu mà không làm mất thông tin quan trọng.
    - SimpleImputer với chiến lược mean thay thế các giá trị thiếu (bao gồm cả null và NaN) bằng trung bình của cột, giữ nguyên ý nghĩa thống kê.
    - Sau khi xử lý, tất cả giá trị thiếu sẽ được điền, đảm bảo dữ liệu không còn khoảng trống.

#### Xóa dữ liệu trùng lặp

In [ ]:
# Kiểm tra và xóa dữ liệu trùng lặp
initial_rows = data.shape[0]
data = data.drop_duplicates()

# Hiển thị số lượng mẫu trước và sau khi xóa trùng lặp
print(f"\nSố lượng mẫu trước khi xóa trùng lặp: {initial_rows}")
print(f"Số lượng mẫu sau khi xóa trùng lặp: {data.shape[0]}")
if initial_rows == data.shape[0]:
    print("Không có dữ liệu trùng lặp trong dataset.")
else:
    print(f"Đã xóa {initial_rows - data.shape[0]} mẫu trùng lặp.")

- Giải thích:

    - drop_duplicates() loại bỏ các hàng có giá trị giống nhau hoàn toàn.
    - Kết quả sẽ cho biết có bao nhiêu mẫu bị trùng lặp (nếu có), giúp dữ liệu trở nên sạch sẽ và đáng tin cậy hơn.

#### Xử lý chuẩn hóa dữ liệu

- Giới thiệu MinMaxScaler (sklearn.preprocessing)
- **MinMaxScaler** là một công cụ trong thư viện **Scikit-learn** dùng để **chuẩn hóa dữ liệu (feature scaling)**.  
- Nguyên lý hoạt động: đưa toàn bộ giá trị của một biến về trong **khoảng [0, 1]** (hoặc một khoảng tùy chọn do người dùng định nghĩa). 
Công thức biến đổi:
$$
X' = \frac{X - X_{min}}{X_{max} - X_{min}}
$$
Trong đó:  
- \(X\) là giá trị gốc,  
- \(X_{min}\) là giá trị nhỏ nhất của cột,  
- \(X_{max}\) là giá trị lớn nhất của cột,  
- \(X'\) là giá trị đã được scale (nằm trong [0,1]). 

In [ ]:
# Chuẩn hóa dữ liệu bằng MinMaxScaler
from sklearn.preprocessing import MinMaxScaler

In [ ]:
# Kiểm tra phạm vi giá trị của các cột
print("\nPhạm vi giá trị của các cột:")
print(data.describe())

scaler = MinMaxScaler()
data_scaled = pd.DataFrame(scaler.fit_transform(data.drop('Outcome', axis=1)), columns=data.drop('Outcome', axis=1).columns)
data_scaled['Outcome'] = data['Outcome']

# Hiển thị 5 dòng đầu của dữ liệu đã chuẩn hóa
print("\n5 dòng đầu của dữ liệu đã chuẩn hóa:")
print(data_scaled.head())

- Giải thích:

    - describe() kiểm tra phạm vi giá trị (min, max) để đánh giá cần chuẩn hóa không. Ví dụ, Age và Pregnancies có thể có giá trị lớn hơn các cột khác.
    - MinMaxScaler chuẩn hóa dữ liệu về khoảng [0, 1], giúp các thuộc tính có cùng thang đo, đặc biệt quan trọng cho mô hình hóa sau này.
    - Dữ liệu chuẩn hóa được lưu vào data_scaled. Có thể gán data = data_scaled nếu muốn sử dụng ngay.

#### Kiểm tra dữ liệu sau khi xử lý

In [ ]:
# Hiển thị thông tin cơ bản sau khi xử lý
print("\nThông tin cơ bản về dữ liệu sau khi xử lý:")
print(data.info())

# Phân bố lớp sau khi xử lý
print("\nPhân bố biến lớp (Outcome) sau khi xử lý:")
print(data['Outcome'].value_counts())

- Giải thích:

    - info() xác nhận không còn giá trị thiếu và kiểu dữ liệu phù hợp.
    - value_counts() kiểm tra phân bố lớp vẫn giữ nguyên (500 mẫu 0 và 268 mẫu 1), đảm bảo không mất dữ liệu quan trọng.

## 4. Phân tích dữ liệu (Analyze Data)

### 4.1 Mô tả vấn đề

- Mục tiêu: 
    - Trong phần này, chúng ta sẽ trực quan hóa dataset Pima Indians Diabetes để khám phá các xu hướng, mối quan hệ giữa các thuộc tính, và nhận xét dựa trên phân tích. Đặc biệt, mình sẽ áp dụng các công thức và phương pháp từ bài báo "Using the ADAP Learning Algorithm" (paper2.pdf), như độ nhạy (sensitivity), độ đặc hiệu (specificity), và đường cong ROC, để đánh giá hiệu suất dự đoán đái tháo đường. Điều này giúp hiểu rõ hơn về dữ liệu và chuẩn bị cho mô hình hóa.

- Dữ liệu: 
    - Dataset đã được làm sạch trong Phần 3 ***(data_scaled)***, với 768 mẫu, 8 thuộc tính y tế, và biến lớp Outcome (0: Không đái tháo đường, 1: Có đái tháo đường).

### 4.2 Thống kê mô tả

#### (1) Hiển thị một số thông tin về dữ liệu
+ Số dòng, số cột của dữ liệu
+ Kiểu dữ liệu của từng cột
+ 5 dòng đầu và 5 dòng cuối của bảng dữ liệu
+ Thông tin chung về dữ liệu

In [ ]:
# shape
print(f'+ Shape: {data_scaled.shape}')
# types
print(f'+ Data Types: \n{data_scaled.dtypes}')
# head, tail
print(f'+ Contents: ')
print('--- First 5 rows ---')
print(data_scaled.head(5))
print('--- Last 5 rows ---')
print(data_scaled.tail(5))
# info
data_scaled.info()

**Nhận xét:**

- Dataset sau khi chuẩn hóa (`data_scaled`) có 768 dòng và 9 cột, không có giá trị thiếu, đảm bảo chất lượng dữ liệu cho phân tích tiếp theo.
- Các thuộc tính số đã được chuẩn hóa về khoảng [0, 1], giúp các đặc trưng có cùng thang đo, thuận lợi cho các thuật toán học máy.
- Kiểu dữ liệu của các cột phù hợp: hầu hết là `float64`, riêng biến mục tiêu `Outcome` là `int64`.
- Phân bố lớp (`Outcome`) vẫn giữ nguyên, đảm bảo không mất dữ liệu quan trọng sau các bước xử lý.
- Việc chuẩn hóa và làm sạch dữ liệu giúp tăng độ tin cậy và hiệu quả cho các bước phân tích, trực quan hóa và xây dựng mô hình dự đoán sau này.

#### (2) Kiểm tra tính toàn vẹn của dữ liệu
1. Có còn giá trị thiếu (NaN, Null) nào trong từng cột không?
2. Có dòng dữ liệu nào bị trùng lặp không?
3. Các giá trị của biến mục tiêu (Outcome) có nằm ngoài tập {0, 1} không?
4. Có cột nào chứa toàn giá trị giống nhau không?
5. Các cột số đã chuẩn hóa có nằm trong khoảng [0, 1] không?

In [ ]:

# Kiểm tra tính toàn vẹn của dữ liệu
print("1. Có còn giá trị thiếu (NaN, Null) nào trong từng cột không?")
print(data_scaled.isnull().sum())

print("\n2. Có dòng dữ liệu nào bị trùng lặp không?")
print(f"Số dòng trùng lặp: {data_scaled.duplicated().sum()}")

print("\n3. Các giá trị của biến mục tiêu (Outcome) có nằm ngoài tập {0, 1} không?")
print(data_scaled['Outcome'].unique())

print("\n4. Có cột nào chứa toàn giá trị giống nhau không?")    
for col in data_scaled.columns:
    if data_scaled[col].nunique() == 1:
        print(f"Cột {col} chỉ có một giá trị duy nhất.")
        
print("\n5. Các cột số đã chuẩn hóa có nằm trong khoảng [0, 1] không?")
for col in data_scaled.columns:
    if col != 'Outcome':
        min_val = data_scaled[col].min()
        max_val = data_scaled[col].max()
        if min_val < 0 or max_val > 1:
            print(f"Cột {col} có giá trị ngoài khoảng [0, 1]: min={min_val}, max={max_val}")

In [ ]:
data.to_csv('output.csv', index=False)

**Nhận xét:**
- Không còn giá trị thiếu (NaN, Null) trong bất kỳ cột nào, đảm bảo dữ liệu đầy đủ cho phân tích.
- Không có dòng dữ liệu nào bị trùng lặp, dữ liệu duy nhất và đáng tin cậy.
- Biến mục tiêu (Outcome) chỉ chứa giá trị 0 và 1, đúng định dạng nhị phân.
- Không có cột nào chứa toàn giá trị giống nhau, đảm bảo các thuộc tính đều có ý nghĩa phân biệt.
- Các cột số đã được chuẩn hóa nằm trong khoảng [0, 1], phù hợp cho các thuật toán học máy và phân tích tiếp theo.

### 4.3 Trực quan hóa dữ liệu

#### (1) Phân bố biến lớp (Outcome)

Giới thiệu thư viện Seaborn và Matplotlib
- **Matplotlib (`matplotlib.pyplot`)**  
  - Là thư viện trực quan hóa dữ liệu cơ bản và mạnh mẽ nhất trong Python.  
  - Cho phép vẽ nhiều loại biểu đồ: line chart, bar chart, scatter plot, histogram, pie chart...  
  - Thường được sử dụng như “nền tảng” cho các thư viện trực quan hóa khác.  


- **Seaborn (`seaborn`)**  
  - Được xây dựng dựa trên Matplotlib, cung cấp **giao diện đơn giản và đẹp mắt** để trực quan hóa dữ liệu thống kê.  
  - Tích hợp tốt với **pandas DataFrame**.  
  - Hỗ trợ nhiều biểu đồ nâng cao: heatmap, pairplot, violin plot, boxplot...  
  - Tự động tối ưu hóa màu sắc và phong cách hiển thị. 

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

In [ ]:
# Phân bố lớp
plt.figure(figsize=(8, 5))
sns.countplot(x='Outcome', data=data_scaled, palette='Set2')
plt.title('Phân bố số lượng bệnh nhân theo trạng thái đái tháo đường')
# Hiển thị thêm con số trên cột
for p in plt.gca().patches:
    plt.gca().annotate(f'{p.get_height()}', (p.get_x() + p.get_width() / 2., p.get_height()), ha='center', va='center', fontsize=11, color='black', xytext=(0, 5), textcoords='offset points')
    # giải thích: p.get_x() + p.get_width() / 2. là vị trí x ở giữa cột, p.get_height() là vị trí y trên đỉnh cột
plt.xlabel('Outcome (0: Không, 1: Có)')
plt.ylabel('Số lượng')
plt.show()

**Nhận xét:** 
- Dữ liệu cho thấy 500 mẫu không đái tháo đường (0) và 268 mẫu có đái tháo đường (1), chiếm tỷ lệ khoảng 65% và 35%. Sự mất cân bằng này cần được xem xét khi xây dựng mô hình, phù hợp với phân bố được ghi nhận trong tài liệu.

#### (2) Phân bố các thuộc tính

Ý nghĩa của việc sử dụng biểu đồ Histogram + KDE để trực quan dữ liệu
- **Histogram (biểu đồ tần suất):**  
  - Cho thấy sự phân bố của dữ liệu theo từng biến (feature).  
  - Giúp ta hiểu được giá trị nào xuất hiện nhiều, giá trị nào ít, có bị lệch (skewness) hay không.  
  - Dễ dàng nhận ra các giá trị bất thường (outliers) hoặc dữ liệu thiếu (ví dụ: nhiều giá trị 0 ở `Insulin`, `SkinThickness`).  

- **KDE (đường mật độ):**  
  - Biểu diễn đường cong mật độ xác suất ước lượng từ histogram.  
  - Làm cho ta thấy rõ hơn **hình dạng phân phối** (chuẩn, lệch phải, lệch trái).  
  - Giúp so sánh mượt mà hơn giữa các nhóm.     

- **Sử dụng `hue='Outcome'`:**  
  - Cho phép **so sánh phân phối** giữa hai nhóm:  
    - `Outcome = 0`: không mắc tiểu đường.  
    - `Outcome = 1`: mắc tiểu đường.  
  - Giúp ta phát hiện biến nào có **khả năng phân biệt mạnh** (ví dụ: Glucose, BMI), và biến nào ít phân biệt (ví dụ: BloodPressure). 

Biểu đồ này giúp:
1. **Hiểu cấu trúc dữ liệu** trước khi làm mô hình.  
2. **Xác định biến quan trọng** (feature importance trực quan).  
3. **Phát hiện dữ liệu bất thường hoặc thiếu**.  
4. Đưa ra định hướng cho **tiền xử lý dữ liệu** (xử lý outlier, missing values, chuẩn hóa). 

In [ ]:
# Vẽ histogram cho tất cả các thuộc tính
plt.figure(figsize=(15, 10))
for i, column in enumerate(data.columns[:-1], 1):  # Loại bỏ 'Outcome' khỏi histogram
    plt.subplot(3, 3, i)
    sns.histplot(data=data, x=column, hue='Outcome', kde=True, palette='deep')
    plt.title(f'Phân bố {column} theo trạng thái đái tháo đường')
    plt.xlabel(column)
    plt.ylabel('Số lượng')
plt.tight_layout()
plt.show()

**Nhận xét:**
- Pregnancies (Số lần mang thai):
    - Phân bố lệch phải, với phần lớn giá trị từ 0-5. Nhóm Outcome = 1 có xu hướng tập trung ở số lần mang thai cao hơn (4-10), gợi ý rằng tiền sử sinh sản nhiều có thể làm tăng nguy cơ, phù hợp với nghiên cứu dịch tễ trong paper3.pdf.

- Glucose (Nồng độ glucose):
    - Đây là biến phân tách rõ ràng nhất:
    - Phân bố gần chuẩn, nhưng nhóm Outcome = 1 có đỉnh rõ ở mức >120 mg/dl, hỗ trợ tiêu chí chẩn đoán WHO (≥200 mg/dl, paper1.pdf).
    - Nhóm tiểu đường có giá trị glucose cao hơn hẳn nhóm không tiểu đường.
    -Đây là biến quan trọng nhất để phân loại.

- BloodPressure (Huyết áp tâm thu):
    -Phân bố tập trung quanh 70-80 mm Hg, không có sự khác biệt rõ rệt giữa hai nhóm, cho thấy huyết áp có thể không phải yếu tố quyết định trong dataset này.

- SkinThickness (Độ dày nếp gấp da):
    - Phân bố lệch phải, nhóm Outcome = 1 có xu hướng cao hơn (20-40 mm), phản ánh mối liên hệ với béo phì, như được đề cập trong paper3.pdf.

- Insulin (Insulin huyết thanh):
    - Phân bố rất lệch, với nhiều giá trị thấp. Nhóm Outcome = 1 có xu hướng cao hơn, nhưng dữ liệu thiếu nhiều (giá trị 0 ban đầu) có thể ảnh hưởng đến phân tích.

- BMI (Chỉ số khối cơ thể):
    - Phân bố tập trung quanh 30-40, với nhóm Outcome = 1 nghiêng về BMI >30 (béo phì), củng cố vai trò của béo phì trong đái tháo đường loại 2 (paper3.pdf).
    - Nhóm tiểu đường có xu hướng BMI cao hơn.
    - Đây là một yếu tố nguy cơ quan trọng.

- DiabetesPedigreeFunction (Hàm phả hệ):
    - Phân bố lệch phải, nhóm Outcome = 1 có giá trị cao hơn (0.4-1.0), cho thấy yếu tố di truyền quan trọng, như phân tích trong paper2.pdf.

- Age (Tuổi):
    - Người tiểu đường có xu hướng lớn tuổi hơn.
    - Yếu tố tuổi tác cũng góp phần tăng nguy cơ.
    - Phân bố lệch phải, nhóm Outcome = 1 tập trung ở tuổi >40, phù hợp với nghiên cứu về lão hóa và nguy cơ đái tháo đường (paper1.pdf).


### 4.4 Trực quan hóa mối quan hệ giữa các thuộc tính

#### (1) Ma trận tương quan

- **Ý nghĩa:**  
  - Ma trận tương quan cho thấy mối quan hệ tuyến tính giữa các cặp biến.  
  - Hệ số tương quan (Pearson):  
    - Gần **1** → tương quan thuận mạnh.  
    - Gần **-1** → tương quan nghịch mạnh.  
    - Gần **0** → ít hoặc không có quan hệ tuyến tính.  
  - Màu sắc trên heatmap:  
    - **Đỏ** = tương quan dương mạnh.  
    - **Xanh** = tương quan yếu hoặc âm.  

In [ ]:
# Ma trận tương quan
plt.figure(figsize=(10, 8))
sns.heatmap(data_scaled.corr(), annot=True, cmap='coolwarm', fmt='.2f')
plt.title('Ma trận tương quan giữa các thuộc tính')
plt.show()

**Nhận xét:**

1. **Mối quan hệ với nhãn `Outcome`:**  
   - `Glucose` có tương quan cao nhất (0.49) → cho thấy nồng độ glucose cao là yếu tố dự báo quan trọng cho bệnh đái tháo đường, phù hợp với kiến thức y học.
   - `BMI` (0.31), `Age` (0.24), `Pregnancies` (0.22), `SkinThickness` (0.22) → có mối quan hệ dương nhưng yếu hơn. cho thấy béo phì, tuổi cao và yếu tố di truyền đều góp phần làm tăng nguy cơ mắc bệnh. 
   - `BloodPressure`, `Insulin`, `DiabetesPedigreeFunction` → tương quan rất thấp (< 0.21).

2. **Mối quan hệ giữa các biến độc lập:**  
   - `SkinThickness` ↔ `BMI` (0.54) → hợp lý, do mỡ dưới da thường liên quan đến BMI.  
   - `Pregnancies` ↔ `Age` (0.54) → càng lớn tuổi thì số lần mang thai càng nhiều.  
   - `Glucose` ↔ `Insulin` (0.42) → đường huyết cao thường đi kèm insulin cao.    

3. **Đánh giá cho mô hình học máy:**  
    - Không có cặp biến nào tương quan quá cao (>0.8) → ít nguy cơ đa cộng tuyến.  
    - Các biến khác nhau đóng góp bổ sung thông tin.  
    - Không có cặp thuộc tính nào có tương quan tuyệt đối (gần 1 hoặc -1), do đó dữ liệu phù hợp cho các mô hình học máy mà không lo trùng lặp thông tin.

#### (2) Hiển thị nhiều tính chất

- ***Phân tích biểu đồ Pairplot***
- Pairplot là một công cụ trực quan hóa mạnh mẽ để hiển thị mối quan hệ giữa tất cả các cặp biến trong dataset. Biểu đồ này tạo ra một ma trận các scatter plot (biểu đồ phân tán) và histogram.
**Cấu trúc biểu đồ:**
- Đường chéo chính: Histogram phân phối của từng biến
- Các ô ngoài đường chéo: Scatter plot giữa hai biến
- Màu sắc: Phân biệt theo nhóm (Outcome: 0 = không đái tháo đường, 1 = có đái tháo đường)
**Mục đích:**
Biểu đồ pairplot này cung cấp một cái nhìn toàn diện về mối quan hệ giữa các yếu tố nguy cơ và tình trạng đái tháo đường. 

In [ ]:
# Vẽ pairplot để phân tích mối quan hệ đa biến
plt.figure(figsize=(12, 10))
sns.pairplot(data, hue='Outcome', palette='deep', diag_kind='hist')
plt.suptitle('Mối quan hệ đa biến giữa các thuộc tính theo trạng thái đái tháo đường', y=1.02)
plt.show()

**Nhận xét:**

1. Glucose (Đường huyết)
    Quan sát: Có sự phân tách rõ ràng giữa 2 nhóm
    Ý nghĩa: Những người có glucose cao (>120-140 mg/dL) có xu hướng mắc đái tháo đường
    Kết luận: Glucose là yếu tố dự đoán mạnh nhất

2. BMI (Chỉ số khối cơ thể)
    Quan sát: Nhóm bị đái tháo đường có BMI cao hơn
    Ý nghĩa: Thể trọng tăng là yếu tố nguy cơ quan trọng
    Kết luận: BMI > 30 tăng nguy cơ đáng kể

3. Age (Tuổi)
    Quan sát: Tần suất bệnh tăng theo tuổi
    Ý nghĩa: Tuổi tác là yếu tố nguy cơ tự nhiên
    Kết luận: Người trên 40 tuổi có nguy cơ cao hơn

4. Pregnancies (Số lần mang thai)
    Quan sát: Có mối liên hệ với bệnh ở phụ nữ
    Ý nghĩa: Đái tháo đường thai kỳ có thể dẫn đến đái tháo đường type 2
    Kết luận: Nhiều lần mang thai tăng nguy cơ

5. DiabetesPedigreeFunction (Yếu tố di truyền)
    Quan sát: Phân bố khác biệt giữa 2 nhóm
    Ý nghĩa: Tiền sử gia đình ảnh hưởng đến nguy cơ
    Kết luận: Di truyền là yếu tố quan trọng

**Mối quan hệ giữa các biến:**
- Tương quan tích cực mạnh:
Glucose - Outcome: Mối liên hệ mạnh nhất
BMI - SkinThickness: Thể hiện tình trạng béo phì
Age - Pregnancies: Tự nhiên theo thời gian

- Tương quan yếu hoặc không đáng kể:
BloodPressure: Không phân biệt rõ ràng giữa 2 nhóm
SkinThickness: Có xu hướng nhưng không mạnh
Insulin: Phân tán nhiều, khó phân biệt

**Ý nghĩa lâm sàng:**
- Chẩn đoán và sàng lọc:
Glucose là chỉ số quan trọng nhất để chẩn đoán
BMI giúp đánh giá nguy cơ
Age và Pregnancies là yếu tố nguy cơ bổ sung

- Dự phòng:
Kiểm soát cân nặng (BMI)
Theo dõi đường huyết định kỳ
Chú ý đặc biệt với phụ nữ có nhiều lần mang thai
Tầm soát sớm ở người có tiền sử gia đình

***Kết luận tổng thể:***
Biểu đồ pairplot cho thấy Glucose, BMI, và Age là ba yếu tố quan trọng nhất trong việc dự đoán đái tháo đường. Sự kết hợp của nhiều yếu tố sẽ cho độ chính xác cao hơn so với chỉ dựa vào một yếu tố đơn lẻ. Điều này hỗ trợ việc phát triển các mô hình machine learning để dự đoán và chẩn đoán bệnh đái tháo đường.

#### (2) Biểu đồ scatter giữa Glucose và BMI (tiêu biểu)

***Phân tích biểu đồ Scatter***
**Ý nghĩa của biểu đồ Scatter Plot**
- Scatter plot (biểu đồ phân tán) hiển thị mối quan hệ giữa hai biến liên tục - trong trường hợp này là Glucose (đường huyết) và BMI (chỉ số khối cơ thể). Màu sắc phân biệt giữa hai nhóm:
    - Màu xanh (0): Không bị đái tháo đường
    - Màu cam (1): Bị đái tháo đường

In [ ]:
# Biểu đồ scatter giữa Glucose và BMI
plt.figure(figsize=(8, 5))
sns.scatterplot(x='Glucose', y='BMI', hue='Outcome', data=data)
plt.title('Mối quan hệ giữa Glucose và BMI theo trạng thái đái tháo đường')
plt.xlabel('Glucose (mg/dl)')
plt.ylabel('BMI (kg/m²)')
plt.show()

**Phân tích chi tiết các vùng trên biểu đồ**
- Vùng 1: Glucose thấp (< 100 mg/dl) - BMI thấp (< 25)
    Đặc điểm: Chủ yếu là điểm màu xanh (không bị bệnh)
    Ý nghĩa: Vùng "an toàn" - ít nguy cơ đái tháo đường nhất
    Khuyến nghị: Duy trì lối sống lành mạnh để giữ ở mức này

- Vùng 2: Glucose cao (> 126 mg/dl) - BMI cao (> 30)
    Đặc điểm: Tập trung nhiều điểm màu cam (bị bệnh)
    Ý nghĩa: Vùng "nguy hiểm" - nguy cơ cao nhất
    Khuyến nghị: Cần can thiệp y tế tích cực, kiểm soát cả đường huyết và cân nặng

- Vùng 3: Glucose trung bình (100-126 mg/dl) - BMI trung bình (25-30)
    Đặc điểm: Hỗn hợp cả hai màu, nhưng có xu hướng tăng màu cam
    Ý nghĩa: Vùng "cảnh báo" - tiền đái tháo đường
    Khuyến nghị: Theo dõi sát, thay đổi lối sống để phòng ngừa

### 4.5 Thuật toán học ADAP để dự đoán sự khởi phát của bệnh đái tháo đường

#### (1) Bối cảnh và Tầm Quan trọng

- Dữ liệu nghiên cứu: Nghiên cứu sử dụng dữ liệu từ dân số Pima Indian, được theo dõi từ năm 1965 bởi National Institute of Diabetes and Digestive and Kidney Diseases (NIDDK). Dân số này có tỷ lệ đái tháo đường cao, đặc biệt là loại không phụ thuộc insulin (NIDDM), do yếu tố di truyền và lối sống.
- Thách thức: Các phương pháp thống kê truyền thống thường thất bại khi dữ liệu có mẫu nhỏ, mối quan hệ hàm phức tạp, hoặc tương tác đa biến không rõ ràng. ADAP nổi bật nhờ khả năng học từ dữ liệu và điều chỉnh dự đoán dựa trên sai số.
- Mục tiêu: Dự đoán sự khởi phát đái tháo đường trong 5 năm dựa trên 8 biến số (số lần mang thai, nồng độ glucose huyết tương, huyết áp, độ dày nếp gấp da, insulin huyết thanh, chỉ số khối cơ thể, hàm số pedigree đái tháo đường, và tuổi).

#### (2) ADAP là gì?

- ADAP là một thuật toán học thích nghi, mô phỏng các thiết bị giống perceptron, được thiết kế để tạo ra các mô hình "phân vùng" đặc biệt cho từng vấn đề cụ thể. Nó học bằng cách điều chỉnh nội bộ khi dự đoán sai, dựa trên sai số giữa giá trị dự đoán và giá trị thực tế.
- Quy trình hoạt động:

    - Đào tạo: Sử dụng 576 trường hợp để ADAP học mẫu từ 8 biến số.
    - Dự đoán: Áp dụng trên 192 trường hợp kiểm tra để dự đoán nguy cơ đái tháo đường trong 5 năm.
    - Điều chỉnh: So sánh kết quả dự đoán với thực tế, điều chỉnh nội bộ để cải thiện độ chính xác.

- Đặc điểm nổi bật: ADAP không yêu cầu giả định về mối quan hệ hàm (khác với hồi quy), phù hợp với dữ liệu y tế phức tạp.

#### (3) Dữ liệu và Biến số Nhập vào: *đã miêu tả ở phần 1 gồm 8 thuộc tính*

- Hàm số DPF được định nghĩa:
$ DPF = \frac{\sum v_i (88 - ADM_i) + 20}{\sum K_j (ACL_j - 14) + 50} $

    - $ v_i $: Tỷ lệ gen chung với họ hàng đã mắc đái tháo đường.
    - $ ADM_i $: Tuổi họ hàng khi được chẩn đoán.
    - $ K_j $: Tỷ lệ gen chung với họ hàng chưa mắc.
    - $ ACL_j $: Tuổi họ hàng ở lần kiểm tra cuối cùng không đái tháo đường.

#### (5) Kết quả và Đánh giá

- Hiệu suất: ADAP đạt độ nhạy (sensitivity) và độ đặc hiệu (specificity) 76% trên 192 trường hợp kiểm tra, với điểm cắt (cutoff) 0.448. Điều này có nghĩa là thuật toán dự đoán đúng 76% trường hợp mắc và không mắc bệnh.
- Đường cong ROC: Được sử dụng để đánh giá hiệu quả phân loại, cho thấy ADAP có khả năng phân biệt tốt giữa các nhóm.
- So sánh: Nghiên cứu đang so sánh ADAP với hồi quy logistic và perceptron tuyến tính để kiểm tra độ ưu việt.

#### (6) Ứng dụng Thực tế


- Y tế dự phòng: ADAP có thể được tích hợp vào hệ thống y tế để sàng lọc sớm nguy cơ đái tháo đường, đặc biệt ở các cộng đồng có nguy cơ cao.
- Cá nhân hóa: Dựa trên dữ liệu cá nhân (8 biến số), bác sĩ có thể đưa ra kế hoạch can thiệp (chế độ ăn, tập luyện) trước khi bệnh phát triển.
- Nghiên cứu thêm: Kết hợp ADAP với dữ liệu di truyền hiện đại (genomics) để nâng cao độ chính xác.

#### (7) Ưu điểm và Hạn chế

**Ưu điểm:**

- Linh hoạt với dữ liệu phức tạp.
- Không yêu cầu giả định thống kê nghiêm ngặt.
- Tiềm năng cải thiện với dữ liệu lớn hơn.

**Hạn chế:**

- Yêu cầu dữ liệu chất lượng cao để đào tạo.
- Hiệu suất 76% vẫn chưa tối ưu, cần cải tiến.
- Chưa được kiểm chứng rộng rãi ngoài dân số Pima.

### 4.6 Mô phỏng và trực quan hóa
**Mô hình:**
- Sử dụng mô hình Logistic Regression để tính và vẽ đường cong ROC, cũng như tính sensitivity và specificity với ngưỡng cắt 0.448 – một thông số được đề cập trong nghiên cứu ADAP.

**Mục tiêu:**
- Mục tiêu là mô phỏng và so sánh hiệu suất dự đoán với kết quả mà ADAP đã đạt được (76% sensitivity và specificity trên 192 mẫu kiểm tra).

**Ý nghĩa:**
- Là một cách để tái hiện và kiểm tra các phương pháp đánh giá từ ADAP, nhưng nó không phải là đánh giá trực tiếp mô hình ADAP.

Chia dữ liệu và huấn luyện mô hình cơ bản

 Giới thiệu `train_test_split` và `LogisticRegression`
 - **`train_test_split` (từ sklearn.model_selection)**  
  - Dùng để **chia dữ liệu** thành 2 (hoặc nhiều) tập:  
    - **Training set**: dùng để huấn luyện mô hình.  
    - **Testing set**: dùng để đánh giá hiệu suất mô hình trên dữ liệu chưa từng thấy.  
  - Giúp kiểm tra khả năng **tổng quát hóa (generalization)** của mô hình, tránh overfitting.  
  - Thông số thường dùng:  
    - `test_size=0.2`: 20% dữ liệu làm tập test.  
    - `random_state=42`: đảm bảo kết quả chia cố định (tái lặp được).  


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

In [ ]:
# Chia dữ liệu thành tập huấn luyện và kiểm tra
X = data_scaled.drop('Outcome', axis=1) # Các biến đầu vào (8 biến)
y = data_scaled['Outcome'] # Biến mục tiêu (Outcome)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42) # Chia 75% huấn luyện, 25% kiểm tra

# Huấn luyện mô hình Logistic Regression
model = LogisticRegression(max_iter=1000) # Tăng max_iter để đảm bảo hội tụ
model.fit(X_train, y_train) # Huấn luyện mô hình

# Dự đoán xác suất
y_prob = model.predict_proba(X_test)[:, 1] # Lấy xác suất của lớp 1 (có đái tháo đường)


**Giải thích:**
- Sử dụng Logistic Regression để mô phỏng, tương tự như cách ADAP tạo giá trị thực từ 0 đến 1, với ngưỡng cắt 0.448 như trong nghiên cứu.

Tính và vẽ đường cong ROC

***Giới thiệu về ROC Curve và AUC***
**ROC Curve (Receiver Operating Characteristic)**
- ROC Curve là một công cụ đánh giá hiệu suất quan trọng trong machine learning, đặc biệt cho các bài toán phân loại nhị phân (binary classification). Đường cong này thể hiện mối quan hệ giữa:
    - Trục Y (True Positive Rate - TPR): Độ nhạy (Sensitivity) = TP / (TP + FN)
    - Trục X (False Positive Rate - FPR): 1 - Độ đặc hiệu (Specificity) = FP / (FP + TN)

**AUC (Area Under the Curve)**
- AUC là diện tích dưới đường cong ROC, có giá trị từ 0 đến 1:
    - AUC = 1.0: Mô hình hoàn hảo
    - AUC = 0.5: Mô hình ngẫu nhiên (không có khả năng phân biệt)
    - AUC < 0.5: Mô hình tệ hơn dự đoán ngẫu nhiên

In [ ]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import roc_curve, auc

In [ ]:
# Tính ROC curve và AUC
fpr, tpr, thresholds = roc_curve(y_test, y_prob)
roc_auc = auc(fpr, tpr)

# Vẽ ROC curve
plt.figure(figsize=(8, 5))
plt.plot(fpr, tpr, color='darkorange', lw=2, label=f'ROC curve (AUC = {roc_auc:.2f})')
plt.plot([0, 1], [0, 1], color='navy', lw=2, linestyle='--', label='Random guess')
plt.xlim([0.0, 1.0])
plt.ylim([0.0, 1.05])
plt.xlabel('False Positive Rate (1 - Specificity)')
plt.ylabel('True Positive Rate (Sensitivity)')
plt.title('Receiver Operating Characteristic (ROC) Curve')
plt.legend(loc="lower right")
plt.show()

**Phân tích kết quả từ biểu đồ (AUC = 0.80)**
Đánh giá chất lượng mô hình:
- Điểm mạnh:
    - AUC = 0.80: Mô hình có hiệu suất tốt (Good performance)
    - Đường cong cong lên phía trên: Cho thấy mô hình tốt hơn dự đoán ngẫu nhiên
    - Tăng nhanh ở đầu: TPR tăng nhanh khi FPR còn thấp

**Ứng dụng trong chẩn đoán đái tháo đường**
- Lợi ích của AUC = 0.80:
    - Sàng lọc cộng đồng:
    - Có thể phát hiện 80% ca bệnh với tỷ lệ dương tính giả chấp nhận được
    - Giảm tải cho hệ thống y tế

- Hỗ trợ bác sĩ:
    - Cung cấp thông tin bổ sung cho quyết định lâm sàng
    - Không thay thế nhưng hỗ trợ chẩn đoán

- Theo dõi nguy cơ:
    - Phân tầng bệnh nhân theo mức độ nguy cơ
    - Lên kế hoạch can thiệp sớm

Tính sensitivity và specificity

In [ ]:
# Tính dự đoán với ngưỡng 0.448 (theo paper2.pdf)
y_pred = (y_prob >= 0.448).astype(int)

# Tính ma trận nhầm lẫn
# Ma trận nhầm lẫn là một bảng tóm tắt hiệu suất của mô hình phân loại.
# Nó hiển thị số lượng dự đoán đúng và sai cho mỗi lớp.
# Từ ma trận nhầm lẫn, ta có thể tính sensitivity và specificity
conf_matrix = confusion_matrix(y_test, y_pred)
tn, fp, fn, tp = conf_matrix.ravel()
# ý nghĩa:
# tn: true negative (dự đoán âm tính đúng)
# fp: false positive (dự đoán dương tính sai)
# fn: false negative (dự đoán âm tính sai)
# tp: true positive (dự đoán dương tính đúng)

# Tính sensitivity và specificity
sensitivity = tp / (tp + fn)  # Độ nhạy: Tỷ lệ dự đoán đúng các ca dương tính
specificity = tn / (tn + fp)  # Độ đặc hiệu: Tỷ lệ dự đoán đúng các ca âm tính

print(f"\nMa trận nhầm lẫn:")
print(conf_matrix)
print(f"Sensitivity (Độ nhạy): {sensitivity:.2f}")
print(f"Specificity (Độ đặc hiệu): {specificity:.2f}")

**Nhận xét:**

- Sensitivity và specificity được tính dựa trên ma trận nhầm lẫn, tương tự phương pháp trong paper2. Kết quả kỳ vọng gần 76%, phù hợp với nghiên cứu ADAP.
- Nếu kết quả thấp hơn, có thể cần điều chỉnh ngưỡng hoặc sử dụng mô hình phức tạp hơn.